Function parameter

zingsheim opened the pull request:
In astutils::isVariableChanged there is a comment "TODO: check if function parameter is const".
I want to put the following changes under discussion by this pull request.

This pull request removes the TODO by having a common function astutils::isVariableChangedByFunctionCall.
The algorithm used is the one from valueflow::bailoutFunctionPar (which handles const parameters correctly).

With this pull request astutils::isVariableChangedByFunctionCall is used in valueflow, too.

(As a minor change it contains a commit which removes the unused ValueKind::Min and ValueKind::Max.)

The question is know if the naming of the function isVariableChangedByFunctionCall is OK and if the location for this function in astutils is OK.

danmar commented on lib/astutils.h:
```diff
@@ -71,8 +72,17 @@ bool isWithoutSideEffects(bool cpp, const Token* tok);
 /** Is scope a return scope (scope will unconditionally return) */
 bool isReturnScope(const Token *endToken);
 
+/** Is varibale changed by function call?
```
varibale

danmar commented on lib/astutils.h:
```diff
@@ -71,8 +72,17 @@ bool isWithoutSideEffects(bool cpp, const Token* tok);
 bool isReturnScope(const Token *endToken);
 
+/** Is varibale changed by function call?
+ * In case the anser of the question is inconcluse, e.g. because the function declaration is not known
```
inconcluse

danmar commented on lib/astutils.h:
```diff
@@ -71,8 +72,17 @@ bool isWithoutSideEffects(bool cpp, const Token* tok);
+ * In case the anser of the question is inconcluse, e.g. because the function declaration is not known
+ * the return value is false and the output parameter inconclusive is set to true
+ *
+ * @param tok           token of varible in function call
```
varible

danmar commented on lib/astutils.h:
```diff
@@ -71,8 +72,17 @@ bool isWithoutSideEffects(bool cpp, const Token* tok);
+ * the return value is false and the output parameter inconclusive is set to true
+ *
+ * @param tok           token of varible in function call
+ * @param inconclusive  output variable which indicates that the answer of the question is inconclusive
```
no settings param

danmar commented on lib/astutils.h:
```diff
@@ -71,8 +72,17 @@ bool isWithoutSideEffects(bool cpp, const Token* tok);
+ * @param tok           token of varible in function call
+ * @param inconclusive  output variable which indicates that the answer of the question is inconclusive
+ */
+bool isVariableChangedByFunctionCall(const Token *tok, const Settings *settings, bool &inconclusive);
```
I would like that inconclusive is a pointer to make it more visible in the function calls that the address is passed to the function and that it's an output parameter.

danmar commented on lib/astutils.cpp:
```diff
@@ -341,7 +342,68 @@ bool isReturnScope(const Token * const endToken)
 }
 
-bool isVariableChanged(const Token *start, const Token *end, const unsigned int varid)
+bool isVariableChangedByFunctionCall(const Token *tok, const Settings *settings, bool & inconclusive)
```
hmm.. it's not necessary with the Settings. The Library is enough. On the other hand it's shorter to access the settings in the function calls. What do you think?

danmar commented on test/testnullpointer.cpp:
```diff
@@ -2503,7 +2503,7 @@ class TestNullPointer : public TestFixture {
     void ticket6505() {
         check("void foo(MythSocket *socket) {\n"
-              "  bool do_write=false;\n"
+              "  bool do_write=0;\n"
```
hmm this looks weird. doesn't "false" get ValueFlow::Value 0?

zingsheim replied:
With "false" the function valueFlowSetConstantValue does not set the value.
For a real run the "false" is probably first simplified to "0". However, in test runs the simplification is not executed before the code is checked.

zingsheim commented:
Sorry for all the typos.

danmar wrote:
> hmm.. it's not necessary with the Settings. The Library is enough. On the other hand it's shorter to access the settings in the function calls. What do you think?

Yes in general I also prefer functions to have minimal requirements on there parameters. I only took the parameter from the already existing "isVariableChanged" function.
However, in future it might be, that some functions in the library are marked as C++ functions only and should not be taken into account for a C source. In this case, the whole setting is needed.

danmar wrote:
> I would like that inconclusive is a pointer to make it more visible in the function calls that the address is passed to the function and that it's an output parameter.

For the out-parameter "inconclusive" I changed it to a pointer and renamed it to "pInconclusive" to indicate that it is a pointer to make the test inside the function "if (pInconclusive)" more clear. The code "if (inconclusive)" would be misleading otherwise in case of "inconclusive" being a pointer.
Do you prefer the pointer to have a default value of "nullptr" to omit the parameter. On the other hand this would make it easy for the caller to accidentally ignore the inconclusive case.

danmar commented:
> For the out-parameter "inconclusive" I changed it to a pointer and renamed it to "pInconclusive" to indicate that it is a pointer to make the test inside the function "if (pInconclusive)" more clear.

In general we don't use "p.." names for pointers. However I am not against it.

I think the code is quite clear:

    if (inconclusive)
        *inconclusive = true;


> Do you prefer the pointer to have a default value of "nullptr" to omit the parameter. On the other hand this would make it easy for the caller to accidentally ignore the inconclusive case.

I would say we don't have a default value. So callers must explicitly ignore it.

zingsheim commented:
Which name do you prefer for ```bool *inconclusive```?
1) inconclusive with a misleading ```if (inconclusive)```.
2) pInconclusive with a hungarian notation notation used nowhere else in the code and the capital ```I``` can easily by misread as small ```l```.
3) inconclusivePtr.

zingsheim commented:
I implemented the ```if (inconclusive != nullptr)``` version.

PS: Please do not edit comments written by me by changing them to your answer.

1. I do not get an email notification for this.
2. It reads as if the questions and opinions in there are my questions and opinions, even though they are yours.

danmar commented:
> PS: Please do not edit comments written by me by changing them to your answer.

Sorry that was by mistake!
